tests/integration/targets/lookup_vault_list/tasks/lookup_vault_list_test.yml

```diff
@@ -75,7 +75,7 @@
     - assert:
         that:
           - test_inexistent is failed
-          - test_inexistent.msg is search("Permission Denied")  # This is the error received when listing a non-existent path
+          - test_inexistent.msg is search("doesn't seem to exist")
```

Comment:
I think we need to reverse this change - this is an example error I get for a list operation on an inexistent path:

```
fatal: [localhost]: FAILED! => {"changed": false, "msg": "Forbidden: Permission Denied to path 'vpn-users_inexistent/metadata'."}
```

Reply:
If you don't have permission to a path in Vault, it will generally give you a 403 instead of a 404, to avoid leaking information about whether something exists at a given path.

But if you have permission (via policy), it gives the 404 correctly, so we do still need to account for that change. The code already did, but that code was never reached in the tests.

The reason the tests also showed 403 is that the policy didn't include permissions for the path being checked. That's one of the things I updated along with this.

Though one thing I noticed is that if the mount point doesn't exist at all, it doesn't matter if the path is in a policy it will still give 403.

So for that reason I also changed the "inexistent" path to use the existing mount point, with a non-existent sub-path, rather than a mount point that didn't exist.

Reply:
Interesting.  When I test it against my vault server, for any non-existent path - whether there is a policy that covers it or not, I get a 403 on list operations.  For get etc. it behaves as you describe, but list appears to give 403 regardless of what the policy says.  e.g. I just created a user with no policy, and tried listing the secrets at the fakesecret mount point:

```
curl -v --header "X-Vault-Token: <redacted>" --request LIST http://127.0.0.1:8200/v1/fakesecret/metadata | jq .
*   Trying 127.0.0.1:8200...
* TCP_NODELAY set
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to 127.0.0.1 (127.0.0.1) port 8200 (#0)
> LIST /v1/fakesecret/metadata HTTP/1.1
> Host: 127.0.0.1:8200
> User-Agent: curl/7.68.0
> Accept: */*
> X-Vault-Token: <redacted>
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 403 Forbidden
< Cache-Control: no-store
< Content-Type: application/json
< Strict-Transport-Security: max-age=31536000; includeSubDomains
< Date: Tue, 17 Jan 2023 15:33:08 GMT
< Content-Length: 60
<
{ [60 bytes data]
100    60  100    60    0     0  20000      0 --:--:-- --:--:-- --:--:-- 20000
* Connection #0 to host 127.0.0.1 left intact
{
  "errors": [
    "1 error occurred:\n\t* permission denied\n\n"
  ]
}
```

When I run the integration tests with the expected error being "doesn't seem to exist" it fails, as the error that is returned by the API call / hvac module is "Permission denied".

```
TASK [lookup_vault_list : Failure expected when inexistent path is listed] *****
fatal: [testhost]: FAILED! => {"msg": "An unhandled exception occurred while templating '{{ lookup('community.hashi_vault.vault_list', vault_kv2_api_list_inexistent_path, **kwargs) }}'. Error was a <class 'ansible.errors.AnsibleError'>, original message: An unhandled exception occurred while running the lookup plugin 'community.hashi_vault.vault_list'. Error was a <class 'ansible.errors.AnsibleError'>, original message: Forbidden: Permission Denied to path 'kv2/metadata/__inexistant'.. Forbidden: Permission Denied to path 'kv2/metadata/__inexistant'."}
...ignoring

TASK [lookup_vault_list : assert] **********************************************
fatal: [testhost]: FAILED! => {
    "assertion": "test_inexistent.msg is search(\"doesn't seem to exist\")",
    "changed": false,
    "evaluated_to": false,
    "msg": "Expected failure but got success or wrong failure message."
}
```

I have removed this section from the policy, and get the same result.

```
  path "{{ vault_kv2_api_list_inexistent_path }}" {
    capabilities = ["list"]
  }
```

So I guess the question is - should we "fix" the Vault API response (i.e. raise an issue there), or modify the tests to expect what is provided (403), or both?  Or maybe I've got the wrong end of the stick!

Reply:
> So I guess the question is - should we "fix" the Vault API response (i.e. raise an issue there), or modify the tests to expect what is provided (403), or both? Or maybe I've got the wrong end of the stick!

Although, the tests run here provide the correct error, I don't really understand why.  I'm running Vault v1.9.2, I wonder if that could be an issue?  I'll upgrade and see...

```
TASK [lookup_vault_list : Failure expected when inexistent path is listed] *****
fatal: [testhost]: FAILED! => {"msg": "An unhandled exception occurred while templating '{{ lookup('community.hashi_vault.vault_list', vault_kv2_api_list_inexistent_path, **kwargs) }}'. Error was a <class 'ansible.errors.AnsibleError'>, original message: An unhandled exception occurred while running the lookup plugin 'community.hashi_vault.vault_list'. Error was a <class 'ansible.errors.AnsibleError'>, original message: The path 'kv2/metadata/__inexistant' doesn't seem to exist.. The path 'kv2/metadata/__inexistant' doesn't seem to exist."}
...ignoring

TASK [lookup_vault_list : assert] **********************************************
ok: [testhost] => {
    "changed": false,
    "msg": "All assertions passed"
}
```

Reply:
Interesting, it could be a version issue. I found this but it seems pretty old; I'd think this was updated before 1.9.x:
- https://github.com/hashicorp/vault/issues/1365

But yeah, since some versions respond with 404, we should account for those too. 

Even if some versions give a 403, that is already covered by the other 403 tests.

Reply:
Just to confirm, in your local test to `http://127.0.0.1:8200/v1/fakesecret/metadata` was `fakesecret` actually created as a mount point? That part is key, it will continue to give 403 if the mount doesn't exist

Reply:
Ah ok, no - it doesn't.  That'll be why, thanks.

However, I just noticed that the test is testing a list operation for 'mount_point/metadata/path' which will never work anyway, even if the mount point and the path exist. (see https://github.com/ansible-collections/community.hashi_vault/pull/343#discussion_r1071924966).

Reply:
OK so whether it was updating or not, I get the (I think) expected behaviour now:

List works for:
* mount_point/metadata/path (where path is a path that contains secrets, not a secret name)
* mount_point/metadata

List gives 403 / permission denied for:
* mount_point_non_existent/metadata
* mount_point/metadata/path_non_existent (only if that is listed in a policy with "deny" capability)

List gives 404 / not found for:
* mount_point/metadata/path_non_existent

Just checking the tests again, and we were testing vault_list for:
* `'{{ vault_kv2_mount_point }}/metadata'` (expecting success)
* `'{{ vault_kv2_mount_point }}/metadata/__inexistent'` (expecting 404)

Which doesn't cover all the scenarios above.  I will make sure we have failure tests for:
* `'{{ vault_kv2_mount_point }}_inexistent/metadata/__inexistent'` (expecting 403)
* `'{{ vault_kv2_mount_point }}/metadata/__inexistent'` with explicit policy denying access (expecting 403)

And also will add a test for:
* `'{{ vault_kv2_mount_point }}/metadata/{{ vault_kv2_path }}'` (expecting success)

I'll try and make the examples clearer to pick up these nuances as well.  Thanks for your patience!

Reply:
Have done this in https://github.com/ansible-collections/community.hashi_vault/pull/343/commits/c294b877a9fe33ca872d5addfff1ef9ce89ef181